File: java/src/main/java/com/example/infrastructure/sqs/SQSSender.java
package com.example.infrastructure.sqs;

import com.amazonaws.services.lambda.runtime.events.SQSEvent;
import com.amazonaws.services.sqs.AmazonSQS;
import com.amazonaws.services.sqs.AmazonSQSClientBuilder;
import com.amazonaws.services.sqs.model.MessageAttributeValue;
import com.amazonaws.services.sqs.model.SendMessageRequest;
import com.example.domain.gateway.MessageSender;

import java.util.HashMap;
import java.util.Map;

public class SQSSender implements MessageSender {

    private final AmazonSQS sqs;
    private final String queueUrl;

    public SQSSender() {
        this.sqs = AmazonSQSClientBuilder.defaultClient();
        this.queueUrl = System.getenv("ORIGINAL_QUEUE_URL");
    }

    @Override
    public void send(String body, Map<String, SQSEvent.MessageAttribute> attributes) {
        Map<String, MessageAttributeValue> msgAttributes = new HashMap<>();

        for (Map.Entry<String, SQSEvent.MessageAttribute> entry : attributes.entrySet()) {
            SQSEvent.MessageAttribute attr = entry.getValue();
            msgAttributes.put(entry.getKey(), new MessageAttributeValue()
                    .withDataType(attr.getDataType())
                    .withStringValue(attr.getStringValue()));
        }

        SendMessageRequest request = new SendMessageRequest()
                .withQueueUrl(queueUrl)
                .withMessageBody(body)
                .withMessageAttributes(msgAttributes);

        sqs.sendMessage(request);
    }
}

File: golang/service/sqs_sender.go
package service

import (
    "os"

    "github.com/aws/aws-sdk-go/aws"
    "github.com/aws/aws-sdk-go/aws/session"
    "github.com/aws/aws-sdk-go/service/sqs"
    "github.com/aws/aws-lambda-go/events"
    "golang_dlq_forwarder/model"
)

type SQSSender struct {
    client   *sqs.SQS
    queueURL string
}

func NewSQSSender() *SQSSender {
    sess := session.Must(session.NewSession())
    client := sqs.New(sess)

    return &SQSSender{
        client:   client,
        queueURL: os.Getenv("ORIGINAL_QUEUE_URL"),
    }
}

func (s *SQSSender) Send(body string, attributes map[string]events.SQSMessageAttribute) error {
    msgAttributes := make(map[string]*sqs.MessageAttributeValue)

    for key, attr := range attributes {
        msgAttributes[key] = &sqs.MessageAttributeValue{
            DataType:    aws.String(attr.DataType),
            StringValue: attr.StringValue,
        }
    }

    _, err := s.client.SendMessage(&sqs.SendMessageInput{
        QueueUrl:          aws.String(s.queueURL),
        MessageBody:       aws.String(body),
        MessageAttributes: msgAttributes,
    })

    return err
}
